### PR TITLE
Шуравина Оксана. Задача 2. Вариант 27. Повышение контраста

### PR DESCRIPTION
**Pull request by shyravina19102004**

### CODE REVIEW

**Review by kobych — CHANGES_REQUESTED** [review]
please add more tests

File: tasks/seq/shuravina_o_contrast/func_tests/main.h
Comment (dasha1112) [inline tasks/seq/shuravina_o_contrast/func_tests/main.h:1]: Please, delete empty file

File: validation()
Comment (dasha1112) [inline validation():1]: Please, delete empty file

Comment (shyravina19102004) [reply validation():1]: fix

File: tasks/mpi/shuravina_o_contrast/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
+#include <gtest/gtest.h>
+
+#include <boost/mpi/communicator.hpp>
+#include <boost/mpi/environment.hpp>
+#include <vector>
+
+#include "mpi/shuravina_o_contrast/include/ops_mpi.hpp"
+
+TEST(shuravina_o_contrast, Test_Contrast_Enhancement_Correct_Output_Size) {
```
[the remark this answers is not in the document]
Comment (shyravina19102004) [reply tasks/mpi/shuravina_o_contrast/func_tests/main.cpp]: fix

**Review by kobych — APPROVED** [review]

**Review by dasha1112 — APPROVED** [review]

**Review by aobolensk — CHANGES_REQUESTED** [review]
You still have too many files

**Comment by m1likus** [comment]
75 changed files....

File: tasks/mpi/chistov_a_sum_of_matrix_elements/func_tests/main.cpp [hunk 2]
```diff
@@ -1,255 +1,255 @@
-#include <gtest/gtest.h>
-
-#include <boost/mpi/communicator.hpp>
-#include <boost/mpi/environment.hpp>
-#include <vector>
-
-#include "mpi/chistov_a_sum_of_matrix_elements/include/ops_mpi.hpp"
```
Comment (aobolensk) [inline tasks/mpi/chistov_a_sum_of_matrix_elements/func_tests/main.cpp:7]: Please, do not modify others' files

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/mpi/chistov_a_sum_of_matrix_elements/func_tests/main.cpp (hunk 2 above)
Comment (shyravina19102004) [reply tasks/mpi/chistov_a_sum_of_matrix_elements/func_tests/main.cpp:7]: there were problems with this file, I returned everything back

Comment (shyravina19102004) [reply tasks/mpi/chistov_a_sum_of_matrix_elements/func_tests/main.cpp]: fix

File: tasks/mpi/shuravina_o_contrast/src/ops_mpi.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  return true;
+}
+
+bool ContrastTaskParallel::post_processing() {
+  internal_order_test();
+  unsigned int delta = 0;
+  if (world.rank() == 0) {
+    delta = taskData->inputs_count[0] / world.size();
+  }
+  broadcast(world, delta, 0);
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_contrast/src/ops_mpi.cpp:71]: Please, move data transfer to ::run() function

File: tasks/mpi/shuravina_o_contrast/src/ops_mpi.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+  if (world.rank() == 0) {
+    auto* tmp_ptr = reinterpret_cast<uint8_t*>(taskData->outputs[0]);
+    for (unsigned i = 0; i < delta; i++) {
+      tmp_ptr[i] = output_[i];
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_contrast/src/ops_mpi.cpp:76]: use std::copy

File: tasks/mpi/shuravina_o_contrast/func_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
+#include <gtest/gtest.h>
+
+#include <boost/mpi/communicator.hpp>
+#include <boost/mpi/environment.hpp>
+#include <vector>
+
+#include "mpi/shuravina_o_contrast/include/ops_mpi.hpp"
+
+TEST(shuravina_o_contrast, Test_Contrast_Enhancement_Single_Element_Input) {
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_contrast/func_tests/main.cpp:9]: Please, add tests on randomly generated images

File: tasks/mpi/shuravina_o_contrast/perf_tests/main.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+TEST(shuravina_o_contrast, Test_Contrast_Enhancement_Single_Element_Input) {
+  boost::mpi::environment env;
+  boost::mpi::communicator world;
+
+  if (world.rank() == 0) {
+    auto taskDataPar = std::make_shared<ppc::core::TaskData>();
+
+    std::vector<uint8_t> input = {50};
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_contrast/perf_tests/main.cpp:36]: Perf tests should follow naming convention (please, read rules) and they should be run on big images

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Review by aobolensk — CHANGES_REQUESTED** [review]
There are two tasks in one PR. This is not allowed

File: tasks/mpi/shuravina_o_contrast/perf_tests/main.cpp (hunk 6 above)
Comment (shyravina19102004) [reply tasks/mpi/shuravina_o_contrast/perf_tests/main.cpp]: fix

File: tasks/mpi/shuravina_o_contrast/func_tests/main.cpp (hunk 5 above)
Comment (shyravina19102004) [reply tasks/mpi/shuravina_o_contrast/func_tests/main.cpp:25]: fix

File: tasks/mpi/shuravina_o_contrast/src/ops_mpi.cpp (hunk 4 above)
Comment (shyravina19102004) [reply tasks/mpi/shuravina_o_contrast/src/ops_mpi.cpp]: fix

File: tasks/mpi/shuravina_o_contrast/func_tests/main.cpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+#include <boost/mpi/collectives.hpp>
+#include <boost/mpi/communicator.hpp>
+#include <boost/mpi/environment.hpp>
+#include <boost/serialization/vector.hpp>
+#include <random>
+#include <vector>
+
+#include "mpi/shuravina_o_contrast/include/ops_mpi.hpp"
+
+std::vector<uint8_t> generateRandomImage(size_t size) {
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_contrast/func_tests/main.cpp:13]: Please, either make it static or wrap with namespace

File: tasks/mpi/shuravina_o_contrast/perf_tests/main.cpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+#include <boost/mpi/communicator.hpp>
+#include <boost/mpi/environment.hpp>
+#include <chrono>
+#include <random>
+#include <vector>
+
+#include "mpi/shuravina_o_contrast/include/ops_mpi.hpp"
+
+TEST(shuravina_o_contrast_perf, Test_Contrast_Enhancement_Large_Image) {
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_contrast/perf_tests/main.cpp:11]: Please, align names of perf tests. There should be 2 perf tests with specific names (for each MPI and sequential version)

File: tasks/mpi/shuravina_o_contrast/src/ops_mpi.cpp [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+namespace shuravina_o_contrast {
+
+bool ContrastTaskParallel::pre_processing() {
+  internal_order_test();
+
+  unsigned int delta = 0;
+  if (world.rank() == 0) {
+    delta = taskData->inputs_count[0] / world.size();
+  }
+  broadcast(world, delta, 0);
```
Comment (aobolensk) [inline tasks/mpi/shuravina_o_contrast/src/ops_mpi.cpp:16]: Please, move all data transfer and computation to ::run() function

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/mpi/shuravina_o_contrast/perf_tests/main.cpp (hunk 8 above)
Comment (shyravina19102004) [reply tasks/mpi/shuravina_o_contrast/perf_tests/main.cpp]: fix

File: tasks/mpi/shuravina_o_contrast/src/ops_mpi.cpp (hunk 9 above)
Comment (shyravina19102004) [reply tasks/mpi/shuravina_o_contrast/src/ops_mpi.cpp:38]: fix

File: tasks/seq/shuravina_o_contrast/perf_tests/main.cpp [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+#include <chrono>
+#include <random>
+#include <vector>
+
+#include "core/perf/include/perf.hpp"
+#include "seq/shuravina_o_contrast/include/ops_seq.hpp"
+
+TEST(shuravina_o_contrast, Test_Contrast_Small) {
+  const int count = 100;
```
Comment (allnes) [inline tasks/seq/shuravina_o_contrast/perf_tests/main.cpp:11]: small count for perf

File: tasks/mpi/shuravina_o_contrast/func_tests/main.cpp [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    std::vector<uint8_t> output(input.size());
+    taskDataPar->outputs.emplace_back(reinterpret_cast<uint8_t*>(output.data()));
+    taskDataPar->outputs_count.emplace_back(output.size());
+
+    shuravina_o_contrast::ContrastTaskParallel contrastTaskParallel(taskDataPar);
+    ASSERT_TRUE(contrastTaskParallel.validation());
+  }
+}
+
+TEST(shuravina_o_contrast, Test_Contrast_Enhancement_Max_Values_Input) {
```
Comment (allnes) [inline tasks/mpi/shuravina_o_contrast/func_tests/main.cpp:67]: please check more cases, e.g. (100x32) (64x23) (27x128)

File: tasks/seq/shuravina_o_contrast/perf_tests/main.cpp (hunk 10 above)
Comment (shyravina19102004) [reply tasks/seq/shuravina_o_contrast/perf_tests/main.cpp]: Fix

File: tasks/mpi/shuravina_o_contrast/func_tests/main.cpp (hunk 11 above)
Comment (shyravina19102004) [reply tasks/mpi/shuravina_o_contrast/func_tests/main.cpp:67]: Fix

**Review by aobolensk — APPROVED** [review]

**Review by allnes — APPROVED** [review]

**Pull request merged**